File: commonlibrary/src/main/java/com/yingwumeijia/commonlibrary/utils/adapter/CommonAdapter.java
package com.yingwumeijia.commonlibrary.utils.adapter;

import android.app.Activity;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Jam on 2016/3/10.
 * 万能适配器
 */
public abstract class CommonAdapter<T> extends BaseAdapter {

    protected LayoutInflater mLayoutInflater;
    protected Context mContext;
    protected ArrayList<T> mDatas;
    private int mItemLayoutId;

    public CommonAdapter(Context mContext, ArrayList<T> mDatas, int mItemLayoutId) {
        this.mLayoutInflater = LayoutInflater.from(mContext);
        this.mContext = mContext;
        this.mDatas = mDatas;
        this.mItemLayoutId = mItemLayoutId;
    }

    @Override
    public int getCount() {
        return mDatas == null ? 0 : mDatas.size();
    }

    @Override
    public T getItem(int position) {
        return mDatas.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        ViewHolder viewHolder = getViewHolder(position, convertView, parent);
        conver(viewHolder, getItem(position), position);
        return viewHolder.getConvertView();
    }

    private ViewHolder getViewHolder(int position, View convertView, ViewGroup parent) {
        return ViewHolder.get(mContext, convertView, parent, mItemLayoutId, position);
    }

    public abstract void conver(ViewHolder helper, T item, int position);

    public List<T> getmDatas() {
        return mDatas;
    }

    public void refresh( ArrayList<T> data){
        if (mDatas == null)mDatas = new ArrayList<T>();
        mDatas.clear();
        mDatas.addAll(data);
        notifyDataSetChanged();
    }

}

File: commonlibrary/src/main/java/com/yingwumeijia/commonlibrary/utils/adapter/CommonAdapter.kt
package com.yingwumeijia.commonlibrary.utils.adapter

import android.app.Activity
import android.content.ClipData
import android.content.Context
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewParent
import android.widget.BaseAdapter

/**
 * Created by jamisonline on 2017/5/25.
 */
abstract class CommonAdapter<T : Any>(var activity: Activity?, var fragment: Fragment?, var data: ArrayList<T>?, var itemLayoutId: Int) : BaseAdapter() {


    override fun getCount(): Int {
        return if (data == null) 0 else data!!.size
    }

    override fun getItem(position: Int): Any? {
        return if (data == null) null else data!![position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View {
        var viewHolder: ViewHolder
        if (activity == null)
            viewHolder = ViewHolder.get(fragment!!, convertView!!, parent!!, itemLayoutId)
        else
            viewHolder = ViewHolder.get(activity!!, convertView!!, parent!!, itemLayoutId)

        conver(viewHolder, getItem(position) as T, position)
        viewHolder.position = position
        return viewHolder.mContentView
    }

    fun getViewHolder(position: Int, convertView: View, parent: ViewGroup): ViewHolder {
        if (activity == null)
            return ViewHolder.get(fragment!!, convertView, parent, itemLayoutId)
        else
            return ViewHolder.get(activity!!, convertView, parent, itemLayoutId)
    }

    abstract fun conver(holder: ViewHolder, item: T, position: Int)


}